### ISSUE TITLE
Completion request can be sent before didOpen and crash language-server

### ISSUE THREAD

**Issue by deathaxe** [tomv564/LSP#145]
While trying to reproduce #131 / #140 I found this possible issue.

If a file is opened for which the server needs to be started first, the first completion request may be sent before the didOpen is. This results in the server to crash and respond with KeyError for each following request.

The most efficient way to reproduce this issue in my case was right after starting ST. If I start working too quickly, I can crash the server for sure. During this period, ST's plugin_host is still quite busy with several background tasks. I enabled debug logging on both pyls and LSP with following results.

What I did to get it fail for sure was (I know it might not be a common use case 😉 ):

1. Start ST
2. quickly after startup start typing `sublime.` and delete it several times to trigger completions.

LSP log

```
LSP: global clients:  ['pyls=True']
LSP: no clients found for window 2
LSP: no clients found for window 2
LSP: starting in C:\Apps\Sublime\Data\Packages
LSP: starting ['pyls', '-v', '--log-file', 'c:/data/pyls.log']
LSP: request 1: initialize 
LSP: client registered for window 2 {'pyls': <LSP.main.Client object at 0x00000196F416A1D0>}
reloading plugin Exalt.impl.plugin
reloading plugin Exalt.impl.validator
reloading plugin Exalt.impl.formatter
LSP: request 2: textDocument/completion    **************  FIRST COMPLETION BEFORE didOpen *****************
LSP: got json:  {"result":{"capabilities":{"codeActionProvider":true,"codeLensProvider":{"resolveProvider":false},"completionProvider":{"resolveProvider":false,"triggerCharacters":["."]},"documentFormattingProvider": ...
LSP: notify: initialized

***
A File Icon: Please restart Sublime Text for the applied icons to take effect ...
***

indexing [job 2]: no files were indexed out of the 2 queued, abandoning crawl
LSP: notify: textDocument/didOpen
LSP: got json:  {"error":{"code":-32000,"message":"Server error","data":{"type":"KeyError","args":["file:///C:/Apps/Sublime/Data/Packages/LSP/main.py"],"message":"'file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'"} ...
LSP: Got error from server:  {'code': -32000, 'message': 'Server error', 'data': {'message': "'file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'", 'args': ['file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'], 'type': 'KeyError'}}
LSP: got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/LSP/main.py","diagnostics":[{"source":"mypy","range":{"start":{"line":2541,"character":9},"end":{"lin ...
LSP: notify: textDocument/didChange
LSP: got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/LSP/main.py","diagnostics":[{"source":"mypy","range":{"start":{"line":2541,"character":9},"end":{"lin ...
LSP: notify: textDocument/didChange
LSP: notify: textDocument/didChange
LSP: got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/LSP/main.py","diagnostics":[{"source":"mypy","range":{"start":{"line":2541,"character":0},"end":{"lin ...
LSP: got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/LSP/main.py","diagnostics":[{"source":"mypy","range":{"start":{"line":2541,"character":9},"end":{"lin ...
```

pyls.log
```
2017-10-02 21:18:55,264 UTC - INFO - pyls.language_server - Starting PythonLanguageServer IO language server
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_completion from <module 'pyls.plugins.completion' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\completion.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_definition from <module 'pyls.plugins.definition' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\definition.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_hover from <module 'pyls.plugins.hover' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\hover.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_references from <module 'pyls.plugins.references' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\references.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_signature_help from <module 'pyls.plugins.signature' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\signature.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin jedi_symbols from <module 'pyls.plugins.symbols' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\symbols.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin mccabe from <module 'pyls.plugins.mccabe_lint' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\mccabe_lint.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin pycodestyle from <module 'pyls.plugins.pycodestyle_lint' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\pycodestyle_lint.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin pydocstyle from <module 'pyls.plugins.pydocstyle_lint' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\pydocstyle_lint.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin pyflakes from <module 'pyls.plugins.pyflakes_lint' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\pyflakes_lint.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin yapf from <module 'pyls.plugins.format' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\format.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Loaded pyls plugin pyls_mypy from <module 'pyls_mypy.plugin' from 'c:\\apps\\python36\\lib\\site-packages\\pyls_mypy\\plugin.py'>
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Updated settings to {'plugins': {'pydocstyle': {'enabled': False}}}
2017-10-02 21:18:55,529 UTC - INFO - pyls.config - Disabled plugins: [<module 'pyls.plugins.pydocstyle_lint' from 'c:\\apps\\python36\\lib\\site-packages\\pyls\\plugins\\pydocstyle_lint.py'>]
2017-10-02 21:18:55,670 UTC - ERROR - pyls.language_server - CAUGHT
Traceback (most recent call last):
  File "c:\apps\python36\lib\site-packages\pyls\language_server.py", line 66, in wrapped
    return func(*args, **kwargs)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 125, in m_text_document__completion
    return self.completions(textDocument['uri'], position)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 59, in completions
    completions = self._hook('pyls_completions', doc_uri, position=position)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 18, in _hook
    doc = self.workspace.get_document(doc_uri) if doc_uri else None
  File "c:\apps\python36\lib\site-packages\pyls\workspace.py", line 48, in get_document
    return self._docs[doc_uri]
KeyError: 'file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'
2017-10-02 21:18:55,685 UTC - ERROR - jsonrpc.manager - API Exception: {'type': 'KeyError', 'args': ('file:///C:/Apps/Sublime/Data/Packages/LSP/main.py',), 'message': "'file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'"}
Traceback (most recent call last):
  File "c:\apps\python36\lib\site-packages\jsonrpc\manager.py", line 108, in _get_responses
    result = method(*request.args, **request.kwargs)
  File "c:\apps\python36\lib\site-packages\pyls\language_server.py", line 66, in wrapped
    return func(*args, **kwargs)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 125, in m_text_document__completion
    return self.completions(textDocument['uri'], position)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 59, in completions
    completions = self._hook('pyls_completions', doc_uri, position=position)
  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 18, in _hook
    doc = self.workspace.get_document(doc_uri) if doc_uri else None
  File "c:\apps\python36\lib\site-packages\pyls\workspace.py", line 48, in get_document
    return self._docs[doc_uri]
KeyError: 'file:///C:/Apps/Sublime/Data/Packages/LSP/main.py'
2017-10-02 21:18:57,607 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:8 W292 no newline at end of file
2017-10-02 21:18:57,607 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:0 E305 expected 2 blank lines after class or function definition, found 0
2017-10-02 21:19:05,576 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:8 W292 no newline at end of file
2017-10-02 21:19:05,576 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:0 E305 expected 2 blank lines after class or function definition, found 0
2017-10-02 21:19:09,311 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:8 W292 no newline at end of file
2017-10-02 21:19:09,311 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:0 E305 expected 2 blank lines after class or function definition, found 0
2017-10-02 21:19:09,811 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:8 W292 no newline at end of file
2017-10-02 21:19:09,811 UTC - INFO - pyls.plugins.pycodestyle_lint - Got pycodestyle error at 2542:0 E305 expected 2 blank lines after class or function definition, found 0
```

It looks like the first `textDocument/completion` is sent before `textDocument/didOpen` which seems to confuse pyls.

Do you have some kind of state machine, which is meant to ensure `textDocument/didOpen` is sent first?

**Comment by tomv564** [comment]
The "document has a client" and "document is opened" states are not connected at the moment, but they could be. If it didn't take deliberate effort to cause this failure, and if `pyls` had no way to recover, we could consider it.

Are you sure `python-language-server` doesn't recover when the `didOpen` notification goes through? Want to look into that?

**Comment by deathaxe** [comment]
pyls keeps throwing KeyError each time a completion is requested. Doesn't recover. The effort needed to cause this issue may depend on the time required to startup and initialize the server, if a document is opened for the very first time. May not be as critical if ST runs for a while compared to startup situation, but I don't know how long other servers need. Maybe java based servers need even more time? Not sure.

I can't say whether pyls is the only one, which can't recover, but from a point of a clean implementation, I think, the didOpen handshake should be the first one.

I was wondering, whether the `DocumentSyncListener` could be used as central object to check and manage the states?

**Comment by tomv564** [comment]
Completions should be disabled until an `on_query_completions` is called that confirms a server has started and offers `completionProvider`.

For the completion request to beat an initial didOpen, it would have to be run between these lines:
https://github.com/tomv564/LSP/blob/e04aa1a0f2767a82d6166924296a3ecb3b9b1ffc/main.py#L1223-L1243

**Comment by deathaxe** [comment]
I honestly find it hard to follow the way each event takes. The whole stuff seems quite scattered over several class and function layers to me. But I guess the issue may be related to the functions being implemented as different threads. Completions are requested by ST synchronous from main thread, while initialization and didOpen/didModify are called asynchronous and all the responses being handled in a third thread as read_stdout is implemented as another one. Furthermore server initialization takes several round trips to complete, which further delays initial didOpen notifications. Server and document states therefore need to be handled by LSP to avoid those race conditions.

**Comment by tomv564** [comment]
The protocol doesn't require `didOpen` for `completion` requests to be completed - see https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#textDocument_didOpen

The completion request must be sent after the `initialize` response has come in, but that is not the issue you are reporting. Can you look into the issue from the `python-language-server` side?

**Comment by deathaxe** [comment]
> The protocol doesn't require didOpen for completion requests to be completed.

I don't want to decline it, but can't read that out of the linked paragraph of the document. Did not read it completely right now, so please excuse some missing knowledge.

While in the first picture here https://github.com/Microsoft/language-server-protocol  the didOpen is the very first and didClose the very last request.

> he completion request must be sent after the initialize response has come in,

From LSP log you'll see this is not the case. The `completion` request is sent before initialization response is received and processed.

Here are the extracted lines

```
LSP: request 1: initialize 
...
LSP: request 2: textDocument/completion  
...
LSP: notify: initialized
```

Following your statement, request 2 must not be sent before `LSP: notify: initialized` 

> Can you look into the issue from the python-language-server side?

I am trying to understand the whole process and find reasons and solutions. As I consider the server expects some of the requests/notifications in a certain order, I started to try to how they are sent as I already found some to seem to be mixed a bit.

**Comment by tomv564** [comment]
Yes, the completion request being sent before `initialized` is a protocol violation, but not the cause of this bug.

Examples of how other language servers solve this:

PHP language server:
https://github.com/felixfbecker/php-language-server/blob/0c399150a3e6349f0d1985c025bd099e3bb15206/src/Server/TextDocument.php#L354-L360

Javascript / Typescript language server: (`overlay` is from didOpen, `files` is disk cache)
https://github.com/sourcegraph/javascript-typescript-langserver/blob/2f97618928eec654c4fd10dbdd19e605e3f67e9f/src/memfs.ts#L162-L172

**Comment by deathaxe** [comment]
> ... but not the cause of this bug.

No it **is** the bug. The server crash is caused by the protocol violation.

Python language-server expects the `uri` sent with the `completion` request to point to a valid document and return the KeyError if not.

https://github.com/palantir/python-language-server/blob/c6da3edead4686732eff02250514dd355e47132d/pyls/python_ls.py#L124-L125

In order to know about the document, a valid `textDocument/didOpen` must have been sent before

https://github.com/palantir/python-language-server/blob/c6da3edead4686732eff02250514dd355e47132d/pyls/python_ls.py#L101-L104

To be able to add the document to the workspace, the `initialization` handshake must have been completed.

https://github.com/palantir/python-language-server/blob/c6da3edead4686732eff02250514dd355e47132d/pyls/python_ls.py#L47-L50


One might argue, whether pyls should handle it more gracefully, but with this protocol violation I would except such an issue to be rejected and the correct initialization order to be implemented by LSP.

The server needs the correct order.

**Comment by tomv564** [comment]
See https://github.com/Microsoft/language-server-protocol/issues/177

**Comment by deathaxe** [comment]
Thank you for the linked info, but unfortunately this issue doesn't answer the question whether it is a client's responsibility to keep order or the servers responsibility to make some assumptions. It sounds as this aspect of usage is just not defined correctly/at all in the specs.

**Comment by tomv564** [comment]
The information you are apparently incapable of reading is:

>  fulfillment of a request should be independent of whether a document is opened or closed

But since requiring all servers to follow the spec might be unproductive, he concedes:

> So it might response with an error if code complete is request on a closed document.

I will change the response handler so the completion state is reset if `pyls` response with an error on a completion request, sometime in the future.

As I alluded to in a comment to your recent pull request, this is exactly the kind of unproductive discussion I want to avoid. I am tired of arguing with you, so I simply won't in the future.
